File: js/model/dialogue/xml.js
/**
 * Shows the xml dialogue and allows the interactions with it
 *
 * @module Dialogue
 */
define(['jquery', 'util/simple-template', 'store/model', 'util/util'],
       function ($, template, store, util) {
    "use strict";

    /** 
     * @class Xml Xml Dialogue class
     * @constructor
     * 
     * @param model {Object} calling model
     */
    function Xml(model) {
        this.model = model // save model

        // find dialogue output
        this.$target = $('#dialogue');

        // set title
        this.$target.find('.modal-title').html('Edit XML');

        this.$body = this.$target.find('.modal-body');
        template.renderTo(this.$body, 'dialogue-xml');

        var self = this;

        // save xml
        this.$target.find('#save').off().click(function () {
            store.updateModel(
                self.model.handle,
                self.$body.find('textarea').val()
            ).then(function () {
                self.$target.modal('hide');
                self.model.reload();
            }).fail(function (e) {
                alert('Could not save XML');
                console.error('XML Saving Failed', e);
            });
        });

        // populate xml
        this.$body.find('textarea').val(util.formatXml(this.model.toXml()));
        
        // show dialogue
        this.$target.modal('show');
    }

    // export
    return Xml;
});

File: src/main/kotlin/model/dialogue/xml.kt
/*
 * Copyright (c) 2017.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with ProcessManager.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package model.dialogue

import BootstrapJQuery
import bootstrap
import jquery.JQuery
import jquery.jQuery
import simpleTemplate
import util.util

/**
 * Kotlin wrapper for the xml class
 */
external interface ModelType {
    var handle: Long
    fun  reload()
    fun  toXml(): String

}


object XmlStatic {
    operator fun invoke(model:ModelType): Xml = Xml(model)
}

class Xml(model: ModelType) {
    var model: ModelType = model
    @JsName("\$target")
    var _target: BootstrapJQuery = jQuery(selector = "#dialogue").bootstrap
    @JsName("\$body")
    var _body: JQuery = _target.find(".modal-body")!!

    init {
        simpleTemplate.renderTo(_body, "dialogue-xml");
        _target.apply {
            find("#save")?.off()?.click {
              store.model.updateModel(model.handle, _body.find("textarea")?.value() as String).then { _ ->
                  _target.modal("hide")
                  model.reload()
              }.fail({ e:Any ->
                  alert("could not save XML")
                  console.error("XML Saving Failed", e)
              })

            }
            modal("show")
        }
        _body.find("textarea")?.value(util.formatXml(this.model.toXml()))
    }
}

val xml:XmlStatic = XmlStatic